aopy/data/bmi3d.py

```diff
@@ -2290,9 +2298,102 @@ def tabulate_stim_data(preproc_dir, subjects, ids, dates, metadata=['stimulation
             continue
 
         # Find laser trial times 
+        idx = np.array([n in exp_metadata.keys() and exp_metadata[n] != '' for n in possible_stim_sites])
+        laser_triggers = np.array(possible_triggers)[idx]
+        laser_sensors = np.array(possible_sensors)[idx]
+        stim_sites = np.array(possible_stim_sites)[idx]
+            
+        for stim_idx in range(len(laser_triggers)):
+            try:
+                trial_times, trial_widths, trial_gains, trial_powers = preproc.bmi3d.get_laser_trial_times(
+                    preproc_dir, subject, te, date, debug=debug, laser_trigger=laser_triggers[stim_idx], 
+                    laser_sensor=laser_sensors[stim_idx], **kwargs)
+            except:
+                print(f"Problem extracting stimulation trials from entry {subject} {date} {te}")
+                traceback.print_exc()
+                continue
+
+            # Tabulate everything together
+            exp = {
+                'subject': subject,
+                'te_id': te, 
+                'date': date, 
+                'trial_time': trial_times,
+                'trial_width': trial_widths, 
+                'trial_gain': trial_gains,
+                'trial_power': trial_powers,
+            }
+
+            # Add requested metadata
+            for key in metadata:
+                if key == 'stimulation_site' and 'qwalor_switch_rdy_dch' in exp_metadata:
+
+                    # Switched laser with multiple stim sites
+                    exp['stimulation_site'] = preproc.bmi3d.get_switched_stimulation_sites(
+                        preproc_dir, subject, te, date, trial_times, debug=debug
+                    )
+
+                elif key == 'stimulation_site' and len(laser_triggers) > 1:
+                    exp['stimulation_site'] = exp_metadata[stim_sites[stim_idx]]
+
+                elif key in exp_metadata:
+                    exp[key] = [exp_metadata[key] for _ in range(len(trial_times))]
+                else:
+                    exp[key] = None
+                    print(f"Entry {subject} {date} {te} does not have metadata {key}.")
+
+
+            # Concatenate with existing dataframe
+            df = pd.concat([df,pd.DataFrame(exp)], ignore_index=True)
+    
+    return df
+
+def tabulate_null_stim_data(preproc_dir, subjects, ids, dates, metadata=[], 
+                            poisson_mu=0.25, refractory_period=0.1, df=None):
+    '''
+    Concatenate null stimulation data by generating poisson-timed stimulation times
+    for the given recordings. Recordings are given as lists of subjects, task entry ids, 
+    and dates. Each list must be the same length. See :func:`~aopy.preproc.utils.generate_poisson_timestamps`
```

Comment:
Is this function normally applied to data with stimulations? Do you need to exclude the the times when the laser is on then?  --- Otherwise maybe it makes more sense to title this something like "tabulate_poisson_time_points"